src/muz/spacer/spacer_global_generalizer.cpp

```diff
@@ -125,39 +127,65 @@ void lemma_global_generalizer::operator()(lemma_ref &lemma) {
     core(lemma);
 }
 
-/// Coerce all uninterpreted constants in \p fml to real
-static void to_real(expr_ref &fml) {
-    ast_manager &m = fml.get_manager();
-    arith_util arith(m);
-    array_util array(m);
-    if (arith.is_numeral(fml) || arith.is_to_real(fml)) return;
-    if (is_uninterp_const(fml) && arith.is_int(fml)) {
-        fml = arith.mk_to_real(fml);
-        return;
+struct to_real_rw_cfg : public default_rewriter_cfg {
+    ast_manager &m;
+    arith_util m_arith;
+    array_util m_array;
+    to_real_rw_cfg(ast_manager &m) : m(m), m_arith(m), m_array(m) {}
+    bool reduce_var(var *v, expr_ref &result, proof_ref &result_proof) {
+        if (m_arith.is_int(v))
+            result = m_arith.mk_to_real(v);
+        else
+            result = v;
+        return true;
     }
-    if (arith.is_arith_expr(fml)) {
-        app *fml_app = to_app(fml);
-        unsigned N = fml_app->get_num_args();
-        expr_ref_vector nw_args(m);
-        expr_ref chld(m);
-        for (unsigned i = 0; i < N; i++) {
-            chld = fml_app->get_arg(i);
-            to_real(chld);
-            nw_args.push_back(chld);
+
+    br_status reduce_app(func_decl *f, unsigned num, expr *const *args,
+                         expr_ref &result, proof_ref &result_pr) {
```

Comment:
should return `BR_FAILED` when no rewrite is needed. I think you are sometimes re-constructing the same expression as input and return `BR_DONE`. This is not the contract of the rewriter.